File: src/ai.js
const database = require('./database');
const { truncate } = require('./utils/ui');

const histories = new Map();
const cooldowns = new Map();
const HISTORY_LIMIT = 12;
const HISTORY_TTL = 30 * 60 * 1000;

const DEFAULT_SYSTEM_PROMPT = [
  'Sen Adash adlı sıcak, samimi ve doğal konuşan bir Discord sohbet arkadaşısın.',
  'Temel amacın bilgi dökmek değil; kullanıcıyla gerçek bir sohbet kurmak, onun üslubuna uyum sağlamak ve gerektiğinde sohbeti nazikçe ilerletmektir.',
  'Türkçe konuş; kısa, canlı ve insani yanıtlar ver. Kullanıcı bir şey paylaşırsa ilgiyle karşılık ver, uygun olduğunda doğal takip soruları sor.',
  'Bilmediğin bir konuda uydurma. Toplu etiket, rol etiketi veya zararlı içerik üretme.'
].join(' ');

function configured() {
  return Boolean(process.env.OPENAI_BASE_URL && process.env.OPENAI_MODEL);
}

function endpoint() {
  const base = process.env.OPENAI_BASE_URL.replace(/\/+$/, '');
  return base.endsWith('/chat/completions') ? base : `${base}/chat/completions`;
}

function safeOutput(value) {
  return String(value)
    .replace(/@everyone/gi, '@\u200beveryone')
    .replace(/@here/gi, '@\u200bhere')
    .replace(/<@&/g, '<@\u200b&')
    .replace(/<@!?\d+>/g, (mention) => mention.replace('@', '@\u200b'));
}

function chunks(value, max = 1900) {
  const result = [];
  let rest = value.trim();
  while (rest.length > max) {
    let index = rest.lastIndexOf('\n', max);
    if (index < max / 2) index = rest.lastIndexOf(' ', max);
    if (index < max / 2) index = max;
    result.push(rest.slice(0, index));
    rest = rest.slice(index).trimStart();
  }
  if (rest) result.push(rest);
  return result;
}

function cleanPrompt(message) {
  return truncate(message.content
    .replace(new RegExp(`<@!?${message.client.user.id}>`, 'g'), '')
    .replace(/<@&\d+>/g, '[rol]')
    .replace(/@everyone|@here/gi, '[toplu etiket]')
    .trim(), 4000);
}

async function requestCompletion(messages) {
  const response = await fetch(endpoint(), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${process.env.OPENAI_API_KEY}` },
    signal: AbortSignal.timeout(Number(process.env.OPENAI_TIMEOUT_MS) || 45_000),
    body: JSON.stringify({
      model: process.env.OPENAI_MODEL,
      messages,
      temperature: Number(process.env.OPENAI_TEMPERATURE) || 0.7,
      max_tokens: Number(process.env.OPENAI_MAX_TOKENS) || 900
    })
  });
  if (!response.ok) throw new Error(`OpenAI uyumlu servis HTTP ${response.status}: ${truncate(await response.text(), 300)}`);
  const data = await response.json();
  const content = data.choices?.[0]?.message?.content;
  if (typeof content !== 'string' || !content.trim()) throw new Error('AI servisi boş yanıt döndürdü.');
  return content.trim();
}

async function handleAiMention(message) {
  if (!configured()) {
    await message.reply({ content: 'Yapay zekâ henüz yapılandırılmamış. `OPENAI_BASE_URL`, `OPENAI_API_KEY` ve `OPENAI_MODEL` gerekli.', allowedMentions: { parse: [], repliedUser: false } });
    return true;
  }
  if (!database.getConfig(message.guild.id, 'ai_enabled', true)) return false;
  const prompt = cleanPrompt(message);
  if (!prompt) {
    await message.reply({ content: 'Bana bir soru da yazmalısın.', allowedMentions: { parse: [], repliedUser: false } });
    return true;
  }
  const cooldownKey = `${message.guild.id}:${message.author.id}`;
  const availableAt = cooldowns.get(cooldownKey) || 0;
  if (availableAt > Date.now()) {
    await message.reply({ content: `Yeni soru için ${Math.ceil((availableAt - Date.now()) / 1000)} saniye bekle.`, allowedMentions: { parse: [], repliedUser: false } });
    return true;
  }
  cooldowns.set(cooldownKey, Date.now() + 10_000);
  setTimeout(() => cooldowns.delete(cooldownKey), 10_000).unref();
  const historyKey = `${message.guild.id}:${message.channel.id}`;
  const stored = histories.get(historyKey);
  const history = stored && Date.now() - stored.updatedAt < HISTORY_TTL ? stored.messages : [];
  const system = database.getConfig(message.guild.id, 'ai_system_prompt', process.env.OPENAI_SYSTEM_PROMPT || DEFAULT_SYSTEM_PROMPT);
  const messages = [{ role: 'system', content: system }, ...history, { role: 'user', content: `${message.author.username}: ${prompt}` }];
  await message.channel.sendTyping();
  const typing = setInterval(() => message.channel.sendTyping().catch(() => null), 7000);
  typing.unref();
  try {
    const raw = await requestCompletion(messages);
    const answer = safeOutput(raw);
    const parts = chunks(answer);
    await message.reply({ content: parts.shift(), allowedMentions: { parse: [], repliedUser: false } });
    for (const part of parts) await message.channel.send({ content: part, allowedMentions: { parse: [] } });
    const updated = [...history, { role: 'user', content: `${message.author.username}: ${prompt}` }, { role: 'assistant', content: answer }].slice(-HISTORY_LIMIT);
    histories.set(historyKey, { messages: updated, updatedAt: Date.now() });
    if (histories.size > 500) histories.delete(histories.keys().next().value);
  } catch (error) {
    console.error('[HATA] yapay zekâ:', error);
    await message.reply({ content: 'Yapay zekâ servisine şu an ulaşılamıyor. URL, model ve API anahtarını kontrol et.', allowedMentions: { parse: [], repliedUser: false } }).catch(() => null);
  } finally {
    clearInterval(typing);
  }
  return true;
}

module.exports = { handleAiMention, requestCompletion, safeOutput, chunks, configured, DEFAULT_SYSTEM_PROMPT };

File: internal/bot/ai.go
package bot

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/adashlabs/adash-bot/internal/config"
	"github.com/adashlabs/adash-bot/internal/database"
	"github.com/bwmarrin/discordgo"
)

const defaultSystemPrompt = "Sen Adash adlı sıcak, samimi ve doğal konuşan bir Discord sohbet arkadaşısın. Temel amacın bilgi dökmek değil; kullanıcıyla gerçek bir sohbet kurmak, onun üslubuna uyum sağlamak ve gerektiğinde sohbeti nazikçe ilerletmektir. Türkçe konuş; kısa, canlı ve insani yanıtlar ver. Bilmediğin bir konuda uydurma. Toplu etiket, rol etiketi veya zararlı içerik üretme."

type chatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}
type chatHistory struct {
	Messages []chatMessage
	Updated  time.Time
}
type aiClient struct {
	cfg       config.Config
	db        *database.DB
	http      *http.Client
	mu        sync.Mutex
	histories map[string]chatHistory
	cooldowns map[string]time.Time
}

func newAI(cfg config.Config, db *database.DB) *aiClient {
	return &aiClient{cfg: cfg, db: db, http: &http.Client{Timeout: cfg.OpenAITimeout}, histories: map[string]chatHistory{}, cooldowns: map[string]time.Time{}}
}
func (a *aiClient) configured() bool { return a.cfg.OpenAIBaseURL != "" && a.cfg.OpenAIModel != "" }
func (a *aiClient) cleanup(now time.Time) {
	a.mu.Lock()
	defer a.mu.Unlock()
	for key, expiry := range a.cooldowns {
		if now.After(expiry) {
			delete(a.cooldowns, key)
		}
	}
	for key, history := range a.histories {
		if now.Sub(history.Updated) > 30*time.Minute {
			delete(a.histories, key)
		}
	}
}

func (a *aiClient) handle(s *discordgo.Session, m *discordgo.MessageCreate) bool {
	if !a.db.ConfigBool(m.GuildID, "ai_enabled", true) {
		return false
	}
	if !a.configured() {
		_, _ = s.ChannelMessageSendReply(m.ChannelID, "Yapay zekâ yapılandırılmamış. OPENAI_BASE_URL, OPENAI_API_KEY ve OPENAI_MODEL gerekli.", m.Reference())
		return true
	}
	prompt := m.Content
	if s.State.User != nil {
		prompt = strings.ReplaceAll(prompt, "<@"+s.State.User.ID+">", "")
		prompt = strings.ReplaceAll(prompt, "<@!"+s.State.User.ID+">", "")
	}
	prompt = regexp.MustCompile(`<@&\d+>`).ReplaceAllString(prompt, "[rol]")
	prompt = strings.TrimSpace(prompt)
	if prompt == "" {
		_, _ = s.ChannelMessageSendReply(m.ChannelID, "Bana bir soru da yazmalısın.", m.Reference())
		return true
	}
	key := m.GuildID + ":" + m.Author.ID
	a.mu.Lock()
	until := a.cooldowns[key]
	if time.Now().Before(until) {
		a.mu.Unlock()
		_, _ = s.ChannelMessageSendReply(m.ChannelID, fmt.Sprintf("Yeni soru için %d saniye bekle.", int(time.Until(until).Seconds())+1), m.Reference())
		return true
	}
	a.cooldowns[key] = time.Now().Add(10 * time.Second)
	historyKey := m.GuildID + ":" + m.ChannelID
	h := a.histories[historyKey]
	if time.Since(h.Updated) > 30*time.Minute {
		h.Messages = nil
	}
	a.mu.Unlock()
	system := a.db.ConfigString(m.GuildID, "ai_system_prompt", valueOr(a.cfg.OpenAISystemPrompt, defaultSystemPrompt))
	messages := append([]chatMessage{{Role: "system", Content: system}}, h.Messages...)
	userText := m.Author.Username + ": " + trunc(prompt, 4000)
	messages = append(messages, chatMessage{Role: "user", Content: userText})
	go func() {
		_ = s.ChannelTyping(m.ChannelID)
		answer, e := a.complete(messages)
		if e != nil {
			_, _ = s.ChannelMessageSendReply(m.ChannelID, "Yapay zekâ servisine şu an ulaşılamıyor. URL, model ve API anahtarını kontrol et.", m.Reference())
			return
		}
		answer = safeAIText(answer)
		parts := chunks(answer, 1900)
		if len(parts) == 0 {
			return
		}
		_, _ = s.ChannelMessageSendReply(m.ChannelID, parts[0], m.Reference())
		for _, x := range parts[1:] {
			_, _ = s.ChannelMessageSend(m.ChannelID, x)
		}
		a.mu.Lock()
		updated := append(h.Messages, chatMessage{Role: "user", Content: userText}, chatMessage{Role: "assistant", Content: answer})
		if len(updated) > 12 {
			updated = updated[len(updated)-12:]
		}
		a.histories[historyKey] = chatHistory{Messages: updated, Updated: time.Now()}
		if len(a.histories) > 500 {
			for k := range a.histories {
				delete(a.histories, k)
				break
			}
		}
		a.mu.Unlock()
	}()
	return true
}
func (a *aiClient) complete(messages []chatMessage) (string, error) {
	endpoint := strings.TrimRight(a.cfg.OpenAIBaseURL, "/")
	if !strings.HasSuffix(endpoint, "/chat/completions") {
		endpoint += "/chat/completions"
	}
	payload := map[string]any{"model": a.cfg.OpenAIModel, "messages": messages, "temperature": a.cfg.OpenAITemperature, "max_tokens": a.cfg.OpenAIMaxTokens}
	raw, _ := json.Marshal(payload)
	ctx, cancel := context.WithTimeout(context.Background(), a.cfg.OpenAITimeout)
	defer cancel()
	req, e := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(raw))
	if e != nil {
		return "", e
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+a.cfg.OpenAIKey)
	res, e := a.http.Do(req)
	if e != nil {
		return "", e
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode > 299 {
		msg, _ := io.ReadAll(io.LimitReader(res.Body, 400))
		return "", fmt.Errorf("AI HTTP %d: %s", res.StatusCode, msg)
	}
	var out struct {
		Choices []struct {
			Message chatMessage `json:"message"`
		} `json:"choices"`
	}
	if e = json.NewDecoder(res.Body).Decode(&out); e != nil {
		return "", e
	}
	if len(out.Choices) == 0 || strings.TrimSpace(out.Choices[0].Message.Content) == "" {
		return "", fmt.Errorf("boş AI yanıtı")
	}
	return strings.TrimSpace(out.Choices[0].Message.Content), nil
}
func safeAIText(value string) string {
	value = safeText(value)
	return regexp.MustCompile(`<@!?\\d+>`).ReplaceAllStringFunc(value, func(mention string) string { return strings.Replace(mention, "@", "@\\u200b", 1) })
}

func chunks(s string, n int) []string {
	var out []string
	s = strings.TrimSpace(s)
	for len([]rune(s)) > n {
		r := []rune(s)
		cut := n
		for cut > n/2 && r[cut] != '\n' && r[cut] != ' ' {
			cut--
		}
		if cut <= n/2 {
			cut = n
		}
		out = append(out, string(r[:cut]))
		s = strings.TrimSpace(string(r[cut:]))
	}
	if s != "" {
		out = append(out, s)
	}
	return out
}
